text/0000-build-write-flag.md

```diff
@@ -0,0 +1,121 @@
+layerize: The process of converting a layer directory to a tarball which will be used to create the final application image during the export phase.
+
+# Motivation
+[motivation]: #motivation
+
+## Why should we do this?
+
+Currently the spec forbids layer modifictions by buildpacks that don't contribute a specific layer. This is not enforced by the lifecycle and leads to cases where subsequent buildpacks can unknowningly modify layer content as a side-effect of executing binaries provided by a build layer. Such unexpected modifications are also really hard to debug and pin-point the case as it can be really difficult to figure out which commands from subsequent buildpacks are causing them.
+
+On the other hand there are valid use cases for build layers that can be used as a collaborative workspace for multiple buildpacks and need to be cached in subsequent runs.
+
+## What use cases does it support?
+
+- Allowing buildpacks to explicitly indicate how the layers provided by them should be used and having this be guaranteed by the lifecycle.
+
+### Example use case for read-only build layers
+
+A build process consists of 2 buildpacks - a python buildpack and a pip buildpack
+
+1. Buildpack #1 provides a python binary that can be used by subsequent buildpacks but is also used in the app image
+2. Buildpack #2 uses the python binary to invoke pip to provide further dependencies.
+
+Unbeknownst to the buildpack author, python has a side-effect on execution where if it's installation directory is writable, it creates new cache objects or updates existing files. As a result, execution of buildpack #2 causes the content of the layers provided by buildpack #1 to change, causing the final layer imageID to change and hampers layer reproducibility and new layers have to be pushed out to the registry.
+
+This could easily be fixed if buildpack #1 only exposed its build layers as read-only to subsequent buildpacks or if it discarded the modifications made by subsequent buildpacks essentially making it "read-only" in the context of the build process.
+
+### Example use case for read and write build layers
+
+There are various use cases when you may want a common workspace that multiple buildpacks can collaborate on during the build process and this workspace needs to be cached, hence it can be a temporary directory.
+
+In such a case a buildpack could provide a layer that is marked as `build-writable` i.e. subsequent buildpacks can modify its contents and the final state of the layer at the end of the build process is what is exported out.
+
+One example might be an Android SDK buildpack that sets up ANDROID_SDK_ROOT (pointing to a layer), which a Gradle buildpack may later write into when gradle runs.
+
+Another example can be a CCache buildpack that sets up `CCACHE_DIR` (pointing to a layer), which a CMake buildpack may use for its build cache.
```

Comment:
I worry about how this interacts with https://github.com/buildpacks/rfcs/pull/151. If for example a CCache buildpack sets up a `CCACHE_DIR` and the CMake buildpack installs things into it, those things should ideally end up in the build BOM. Should the downstream buildpack then edit the layer content metadata of the ccache layer. Just having to figure out the path to the layer content metadata seems to break some important abstractions.